lib/opal/rewriters/returnable_logic.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+require 'opal/rewriters/base'
+
+module Opal
+  module Rewriters
+    class ReturnableLogic < Base
+      def next_tmp
+        @counter ||= 0
+        @counter += 1
+        "$ret_or_#{@counter}"
+      end
+
+      def reset_tmp_counter!
+        @counter = nil
+      end
+
+      def on_or(node)
+        lhs, rhs = *node.children
+        lhs_tmp = next_tmp
+
+        node.updated(:if, [s(:lvasgn, lhs_tmp, process(lhs)), s(:js_tmp, lhs_tmp), process(rhs)])
```

Comment:
oh I have no idea why Opal got it now. Looks like are great idea to me, and I believe we could implement it much, much earlier.

IIRC sorbet has a very similar idea of "ast desugaring". Basically they rewrite AST to have as few nodes as possible by converting them into other "trivial" nodes that they support by default. For example, they rewrite `case` into a chain of `if`s

Reply:
This is great! https://github.com/sorbet/sorbet/blob/master/ast/desugar/Desugar.cc
I'll see if I can back port other stuff from it, and get less and more robust code 👍